File: Airship/AirshipCore/Source/Internal/UALocaleManager.m
/* Copyright Airship and Contributors */

#import "UALocaleManager+Internal.h"

#if __has_include("AirshipCore/AirshipCore-Swift.h")
#import <AirshipCore/AirshipCore-Swift.h>
#elif __has_include("Airship/Airship-Swift.h")
#import <Airship/Airship-Swift.h>
#endif

NSString *const UALocaleUpdatedEventLocaleKey = @"com.urbanairship.locale.locale";
NSString *const UALocaleUpdatedEvent = @"com.urbanairship.locale.locale_updated";

@interface UALocaleManager()
@property (nonatomic, strong) UAPreferenceDataStore *dataStore;
@property (nonatomic, strong) NSNotificationCenter *notificationCenter;
@end

@implementation UALocaleManager

+ (instancetype)localeManagerWithDataStore:(UAPreferenceDataStore *)dataStore {
    return [[UALocaleManager alloc] initWithDataStore:dataStore notificationCenter:[NSNotificationCenter defaultCenter]];
}

- (instancetype)initWithDataStore:(UAPreferenceDataStore *)dataStore notificationCenter:(NSNotificationCenter *)notificationCenter {
    self = [super init];
    if (self) {
        self.dataStore = dataStore;
        self.notificationCenter = notificationCenter;
    }
    return self;
}

- (void)setCurrentLocale:(NSLocale *)currentLocale {
    if ([self.currentLocale isEqual:currentLocale]) {
        return;
    }

    if (!currentLocale) {
        [self clearLocale];
        return;
    }
    NSData *encodedLocale = [NSKeyedArchiver archivedDataWithRootObject:currentLocale];
    [self.dataStore setObject:encodedLocale forKey:UALocaleUpdatedEventLocaleKey];

    NSDictionary *localeDictionary = [NSDictionary dictionaryWithObject:currentLocale forKey:UALocaleUpdatedEventLocaleKey];
    [self.notificationCenter postNotificationName:UALocaleUpdatedEvent object:localeDictionary];
}

- (NSLocale *)currentLocale {
    NSData *encodedLocale = [self.dataStore objectForKey:UALocaleUpdatedEventLocaleKey];
    if (encodedLocale) {
        return [NSKeyedUnarchiver unarchiveObjectWithData:encodedLocale];
    } else {
        return [NSLocale autoupdatingCurrentLocale];
    }
}

- (void)clearLocale {
    [self.dataStore removeObjectForKey:UALocaleUpdatedEventLocaleKey];

    NSLocale *defaultLocale = [NSLocale autoupdatingCurrentLocale];
    NSDictionary *localeDictionary = [NSDictionary dictionaryWithObject:defaultLocale forKey:UALocaleUpdatedEventLocaleKey];
    [self.notificationCenter postNotificationName:UALocaleUpdatedEvent object:localeDictionary];
}


@end

File: Airship/AirshipCore/Source/UALocaleManager.swift
/* Copyright Airship and Contributors */

/**
 * Airship locale manager.
 */
@objc
public class UALocaleManager : NSObject {

    private static let storeKey = "com.urbanairship.locale.locale"

    @objc
    public static let localeUpdatedEvent = NSNotification.Name("com.urbanairship.locale.locale_updated")

    @objc
    public static let localeEventKey = "locale"

    private var dataStore: UAPreferenceDataStore
    private var notificationCenter: NotificationCenter

    /**
     * The current locale used by Airship. Defaults to `autoupdatingCurrent`.
     */
    @objc
    public var currentLocale : Locale {
        get {
            if let encodedLocale = dataStore.object(forKey: UALocaleManager.storeKey) as? Data {
                if let locale = NSKeyedUnarchiver.unarchiveObject(with: encodedLocale) as? Locale {
                    return locale
                }
            }
            return Locale.autoupdatingCurrent
        }
        set {
            let encodedLocale: Data = NSKeyedArchiver.archivedData(withRootObject: newValue)
            dataStore.setValue(encodedLocale, forKey: UALocaleManager.storeKey)
            notificationCenter.post(name: UALocaleManager.localeUpdatedEvent, object:[UALocaleManager.localeEventKey: newValue])
        }
    }

    /**
     * @note For internal use only. :nodoc:
     */
    @objc
    public convenience init(dataStore: UAPreferenceDataStore) {
        self.init(dataStore: dataStore, notificationCenter: NotificationCenter.default)
    }

    /**
     * @note For internal use only. :nodoc:
     */ 
    @objc
    public init(dataStore: UAPreferenceDataStore, notificationCenter: NotificationCenter) {
        self.dataStore = dataStore
        self.notificationCenter = notificationCenter
        super.init()
    }

    /**
     * Resets the current locale.
     */
    @objc
    public func clearLocale() {
        dataStore.removeObject(forKey: UALocaleManager.storeKey)
        notificationCenter.post(name: UALocaleManager.localeUpdatedEvent, object:[UALocaleManager.localeEventKey: self.currentLocale])
    }
}
